uxarray/grid.py

```diff
@@ -0,0 +1,244 @@
+"""uxarray grid module."""
+import os
+import xarray as xr
+import numpy as np
+from warnings import warn
+from pathlib import PurePath
+from datetime import datetime
+
+# reader and writer imports
+from ._exodus import read_exodus, write_exodus
+from ._ugrid import read_ugrid, write_ugrid
+from ._shapefile import read_shpfile
+from ._scrip import read_scrip
+from .helpers import determine_file_type
+
+
+class Grid:
+    """The Uxarray Grid object class that describes an unstructured grid.
+
+    Examples
+    ----------
+    Open an exodus file with Uxarray Grid object
+
+    >>> mesh = ux.Grid("filename.g")
+
+    Save as ugrid file
+
+    >>> mesh.saveas("outfile.ug")
+    """
+
+    def __init__(self,
+                 *args,
+                 filepath=None,
+                 gridspec=None,
+                 vertices=None,
+                 concave=None,
+                 islatlon=None,
+                 mesh_filetype=None,
+                 **kwargs):
+        """Initialize grid variables, decide if loading happens via file, verts
+        or gridspec If loading from file, initialization happens via the
+        specified file.
+
+        # TODO: Add or remove new Args/kwargs below as this develops further
+
+        Parameters
+        ----------
+        args : string, optional
+            - Input file name with extension or
+            - Vertex coordinates that form one face.
+        islatlon : bool, optional
+            Specify if the grid is lat/lon based:
+        concave: bool, optional
+            Specify if this grid has concave elements (internal checks for this are possible)
+        gridspec: bool, optional
+            Specifies gridspec
+        mesh_filetype: string, optional
+            Specify the mesh file type, eg. exo, ugrid, shp etc
+        kwargs : dict, optional
+            A dict initializing specific variables as stated above
+            - example: kwargs = {"concave" : True, "islatlon" : True"}
+
+        Raises
+        ------
+            RuntimeError: File not found
+        """
+        # TODO: fix when adding/exercising gridspec
+        self.filepath = None
+        self.gridspec = None
+        self.vertices = None
+        self.islatlon = None
+        self.concave = None
+        self.mesh_filetype = None
+
+        # internal uxarray representation of mesh stored in internal object in_ds
+        self.in_ds = xr.Dataset()
+
+        # Parse the keyword arguments and initialize class attributes
+        for key, value in kwargs.items():
+            if key == "latlon":
+                self.islatlon = value
+            elif key == "gridspec":
+                self.gridspec = value
+
+        # determine initialization type - string signifies a file, numpy array signifies a list of verts
```

Comment:
> > Should we also expand the testing done in this PR or create a secondary PR for that purpose?
> 
> I am on board for expanding testing within this PR. There isn't a specific test for just the exodus function yet (which I think it could be argued that if current tests work, then _exodus.py is working), but I feel there should be more edge case testing for _exodus.py and grid.py overall. For me, I think when we have other file types added there will be overlaps with variable naming, or users may specify outfiles with unique suffixes (e.g. ".nc", "ugrid", no suffix, etc) and that may start to confuse the Grid class, so while we could wait to see on the tests, it may be better to just address things now instead of later. But, I can also, again, see arguments against addressing these things in this specific PR/at this point in the development process. So, if these aren't necessarily shared concerns to be addressed immediately, I'm fine with expanding tests in a new PR after this one is merged.

I'm envisioning test_grid.py contain all reader/writer tests. It'd be easier and more manageable if we have test_exodus.py and so on, is that in-accordance with the conventions? @erogluorhan

Reply:
Yeah, I see what you mean, and I can also be fine with that! Even though the convention is that we should unit-test each Python module in a separate test file dedicated to it, I don't have a strong feeling when it comes to such a case where grid.py module uses helper functions from other "_..." modules.